technical-docs/K8SSANDRA-OPERATOR-DESIGN.md

```diff
@@ -0,0 +1,410 @@
+# K8ssandra Operator Design
+
+Author: @jsanda
+
+## Overview
+
+GH issue: https://github.com/k8ssandra/k8ssandra/issues/683
+
+This is a design document for an operator for k8ssandra. For background see 
+https://github.com/k8ssandra/k8ssandra/issues/485.
+
+Much of the behavior that is implemented in the Helm charts will be reimplemented in various controllers. The role of the Helm charts changes to focus primarily on installation and configuration of operators.
+
+While we will continue to use and support Helm, it will not be required for installing k8ssandra from a purely technical perspective. Users can install manifests with kubectl apply for example. We can also provide support for kustomize.
+
+## CustomResourceDefinitions
+
+We will utilize a number of CRDs, some of which are developed and maintained by other projects.
+
+There will be a separate controller for each CRD.
+
+The goal of this document is not to be entirely prescriptive for each CRD. Rather, the intent is to give a general idea of how things will look and fit together. 
+
+### K8ssandraCluster
+
+This is the primary object which users will manage. The controller will be responsible for creating other constituent objects.
+
+It will look something like this:
+
+```
+apiVersion: k8ssandra.io/v1alpha1
+kind: K8ssandraCluster
+metadata:
+  name: k8ssandra
+spec:
+  cassandra:
+    clusterName: k8ssandra
+    auth:
+      enabled: true
+    datacenters:
+    - name: dc1
+      size: 3
+  reaper:
+    enabled: true
+    # optionally set and override defaults for Reaper    
+  stargate: 
+    enabled: true  
+    # optionally set and override defaults for Stargate 
+  monitoring:
```

Comment:
Newest operator ships with kube-auth-proxy also (as optional dependency) that allows auth to restrict metrics endpoints. Do we wish to support those features also? Also, operator-sdk generates the ServiceMonitors and metrics configuration automatically these days. So we might want to relook at using kube-prometheus-stack (including the fact that Grafana is problematic)

Reply:
Am I correct in thinking these ServiceMonitors are for the operators themselves?